akka-stream/src/main/scala/akka/stream/impl/FanIn.scala

```diff
@@ -183,13 +204,19 @@ private[akka] object FanIn {
     }
 
     def inputsAvailableFor(id: Int) = new TransferState {
-      override def isCompleted: Boolean = depleted(id) || cancelled(id)
-      override def isReady: Boolean = pending(id)
+      override def isCompleted: Boolean = {
+        val state = states(id)
+        has(state, Depleted) || has(state, Cancelled)
+      }
```

Comment:
Are you certain that you are not over-optimizing prematurely? I’d rather see JMH proof that this helps before making the code more brittle. The method I’m talking about would fetch a byte from an array and do a bit op on it, and it would be highly likely that the CPU would even elide the load for the second invocation, since it knows that the value does not change (no `@volatile` in sight). If this really is a hot path then the right things should happen.